src/librbd/image_copy/ImageCopyRequest.cc

```diff
@@ -0,0 +1,189 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "ImageCopyRequest.h"
+#include "ObjectCopyRequest.h"
+#include "common/errno.h"
+#include "librbd/Utils.h"
+#include "tools/rbd_mirror/ProgressContext.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::image_copy::ImageCopyRequest: " \
+                           << this << " " << __func__ << ": "
+
+namespace librbd {
+namespace image_copy {
+
+using librbd::util::create_context_callback;
+using librbd::util::unique_lock_name;
+
+template <typename I>
+ImageCopyRequest<I>::ImageCopyRequest(I *src_image_ctx, I *dst_image_ctx,
+                                      const SnapSeqs &snap_seqs,
+                                      ProgressContext *prog_ctx,
+                                      Context *on_finish)
+  : RefCountedObject(dst_image_ctx->cct, 1), m_src_image_ctx(src_image_ctx),
+    m_dst_image_ctx(dst_image_ctx), m_snap_seqs(snap_seqs),
+    m_prog_ctx(prog_ctx), m_on_finish(on_finish), m_cct(dst_image_ctx->cct),
+    m_lock(unique_lock_name("ImageCopyRequest::m_lock", this)) {
+}
+
+template <typename I>
+void ImageCopyRequest<I>::send() {
+  int r = compute_snap_map();
+  if (r < 0) {
+    finish(r);
+    return;
+  }
+
+  send_object_copies();
+}
+
+template <typename I>
+void ImageCopyRequest<I>::cancel() {
+  Mutex::Locker locker(m_lock);
+
+  ldout(m_cct, 20) << dendl;
+  m_canceled = true;
+}
+
+template <typename I>
+void ImageCopyRequest<I>::send_object_copies() {
+  m_object_no = 0;
+  uint64_t end_src_object_no;
+  {
+    RWLock::RLocker snap_locker(m_src_image_ctx->snap_lock);
+    end_src_object_no =  m_src_image_ctx->get_object_count(CEPH_NOSNAP);
+    for (auto snap_id : m_src_image_ctx->snaps) {
+      end_src_object_no = std::max(end_src_object_no,
+                                   m_src_image_ctx->get_object_count(snap_id));
+    }
+  }
+
+  // Recalculate for destination
```

Comment:
Why would the destination size be different on a copy?

Reply:
Not the destination image size but total number of objects (`m_end_object_no`) may be different if src and dst object sizes are different.

Reply:
Good point, but the way that ```ObjectCopyRequest``` is currently designed, it assumes that the source and destination have matching layouts (order and striping v2). If we want to support overriding those settings on the destination, which I think is important, that class needs to be redesigned so that you loop over all destination objects and then the ```ObjectCopyRequest``` state machine loads diffs for all source objects that overlap w/ the destination object's layout, computes the necessary ops over the overlap, and does the single destination write per snapshot. 

Given that, I think you would drop this logic and change the loop above to iterate over the destination image instead of the source image.

Reply:
I suppose it is exactly how it works now? I.e. here we calculate number of destination objects and then for every destination object call `ObjectCopyRequest`. In my modified implementation it expects destination object number as a param, and then it maps dst object to src objects, loads diffs, etc...

But I forgot about striping v2 feature. I need to refresh memory about how it works, but right now I think the code will not work correctly in general case if striping is enabled.

Reply:
I missed that you adjusted it to handle that case. All object and offset calculation should use the Striper logic instead of attempting to re-implement it all over the place. You would use the striper to calculate the image extents covered by the destination object and then for each image extent you would map that back to source object extents.